### ISSUE TITLE
generate requestCode is repeat

### ISSUE THREAD

**Issue by xieyingwu** [permissions-dispatcher/PermissionsDispatcher#519]
A activity request LOCATION permission, B activity extands A ,B request CAMERA permission,but generate DispatcherClass  requestCode is same, is all equal 0, how to manual to set request code.??????

**Comment by xieyingwu** [comment]
@eyedol @punchdrunker @snkashis @egor-n @hotchemi

**Comment by xieyingwu** [comment]
i found the real point, B extands A,C extands B, B not request permission,but A request,C request, i found generate DispatcherClass requestCode is same, is all equal 0, how to manual to set request code.??????

**Comment by hotchemi** [comment]
Sorry I don't understand your saying. please paste sample if you need a help

**Comment by mannodermaus** [comment]
From what I understand, this is @xieyingwu's setup:

```kotlin
@RuntimePermissions
open class ActivityA : AppCompatActivity {
  @NeedsPermission(ACCESS_FINE_LOCATION)
  fun requestLocation() {
  }
  // ...
}

```

```kotlin
open class ActivityB : ActivityA
```

```kotlin
@RuntimePermissions
class ActivityC : ActivityB {
  fun requestCamera() {
  }
  // ...
}
```

After compilation, `ActivityAPermissionsDispatcher` and `ActivityCPermissionsDispatcher` generate `requestCode=0` for different permissions. I suppose that this scenario could cause confusion, depending on which generated class checks first, since both `ACCESS_FINE_LOCATION` and `CAMERA` could match inside `onRequestPermissionsResult()`.

**Comment by hotchemi** [comment]
(Sorry for the late)

Umm yea weird.. 🤔 We initialize `RequestCodeProvider` in each processor so there's a case that processor module can be generated multiple times(using multi module?) 🤔 

https://github.com/permissions-dispatcher/PermissionsDispatcher/blob/e6759e1976fff9bbe1aaf787b64ced62d8ecae85/processor/src/main/kotlin/permissions/dispatcher/processor/PermissionsProcessor.kt#L56

One easiest approach that I've come up with is to make RequestCodeProvider really a singleton Object 🤔 

```kt
object RequestCodeProvider {

    private val currentCode = AtomicInteger(0)

    /**
     * Obtains the next unique request code.
     * This method atomically increments the value
     * returned upon the next invocation.
     */
    fun nextRequestCode(): Int = currentCode.andIncrement
}
```

@xieyingwu can we get more detail if possible?

**Comment by mannodermaus** [comment]
The multi-process idea might be interesting... @xieyingwu are the `Activity` classes in different modules?

**Comment by xieyingwu** [comment]
yes, the activity is in different modules, in the same module, A extands B is right, so i adjust provide the params to define the requestCode by user. @mannodermaus @hotchemi

**Comment by xieyingwu** [comment]
I had a idea, the requestCode can be save in one properties file, to define make the generate requestCode is dif

**Comment by hotchemi** [comment]
I also came up with just using random int as initial value.

```kt
object RequestCodeProvider {

    private val currentCode = AtomicInteger(Random().nextInt(1000))

    /**
     * Obtains the next unique request code.
     * This method atomically increments the value
     * returned upon the next invocation.
     */
    fun nextRequestCode(): Int = currentCode.andIncrement
}
```

**Comment by mannodermaus** [comment]
Hmm, using a random seed is likely to break [deterministic, reproducible builds](https://github.com/permissions-dispatcher/PermissionsDispatcher/pull/342) though... also, that could make our code generation tests pretty difficult to handle.

It definitely looks like in a multi-module environment, different instances of the annotation processor are used to process a set of classes, and since each of them uses a new `RequestCodeProvider`, the generated codes might overlap. I'm thinking if there is some kind of way to share annotation processor data between modules, but I've got nothing so far...

**Comment by hotchemi** [comment]
Oh true.. 😮 
Umm I suppose, even each module has it's own processor instance its Java process is the same..then https://github.com/permissions-dispatcher/PermissionsDispatcher/issues/519#issuecomment-418623180 would work..may be.. 😇

**Comment by mannodermaus** [comment]
It's worth a try!

**Comment by hotchemi** [comment]
We've tried on https://github.com/permissions-dispatcher/PermissionsDispatcher/pull/529 but after some considerations it's not an optimal idea 🤔
Another way is to add a new argument to RuntimePermission annotation like below. If you don't specify value for `initialRequestCode` result would be same as current one.

```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.CLASS)
public @interface RuntimePermissions {
  int initialRequestCode() default 0;
}
```

But this is just for a corner case and I'm not sure it would be beneficial to users..🤔

@xieyingwu Can I ask why you need such a complex hierarchy? I wonder you kind of abuse inheritance and solve the problem easily by sorting out your module structure. Do you really need three stages inheritance hierarchy and put activities in different modules?

**Comment by hotchemi** [comment]
ping @xieyingwu
